jbmc/src/java_bytecode/convert_java_nondet.cpp

```diff
@@ -29,83 +29,118 @@ Author: Reuben Thomas, reuben.thomas@diffblue.com
 /// \param target: One of the steps in that goto program.
 /// \param symbol_table: The global symbol table.
 /// \param message_handler: Handles logging.
-/// \param max_nondet_array_length: Maximum size of new nondet arrays.
-/// \return The next instruction to process with this function.
-static goto_programt::targett insert_nondet_init_code(
+/// \param object_factory_parameters: Parameters for the generation of nondet
+///   objects.
+/// \param mode: Language mode
+/// \return The next instruction to process with this function and a boolean
+///   indicating whether any changes were made to the goto program.
+static std::pair<goto_programt::targett, bool> insert_nondet_init_code(
   goto_programt &goto_program,
   const goto_programt::targett &target,
   symbol_table_baset &symbol_table,
   message_handlert &message_handler,
   object_factory_parameterst object_factory_parameters,
   const irep_idt &mode)
 {
-  // Return if the instruction isn't an assignment
-  const auto next_instr=std::next(target);
-  if(!target->is_assign())
-  {
-    return next_instr;
-  }
+  const auto next_instr = std::next(target);
+  bool changed = false;
 
-  // Return if the rhs of the assignment isn't a side effect expression
-  const auto &assign=to_code_assign(target->code);
-  if(assign.rhs().id()!=ID_side_effect)
+  // We only expect to find nondets in the rhs of an assignments or in return
+  // statements
+  for(exprt &op : target->code.operands())
```

Comment:
These are instructions right? Consider `instruction` rather than `op`

Reply:
`target->code` is a `codet`. Its operands are expressions. An assignment will have two operands, the lhs and the rhs, and a return statement will have one operand, the thing being returned.